### PR TITLE
Add linear models

### PR DESCRIPTION
**Pull request by zfrangella**
- Adds core linear models from scikit-learn: Linear, Huber, Logistic, Multinomial, Poisson, Gamma, Inverse Gaussian, Compound Poisson Gamma.
- Implementations support in-memory and out-of-memory data.
- Linear models have core api methods like predict, score, decision_function (classifiers only) etc.
- Adds comprehensive tests for the implementations across input modalities. Also tests CI with proximal gradient to ensure these methods work with existing optimizers. 
- Fixes an issue with Polyhedron with device management with lower and upper.

### CODE REVIEW

File: rlaopt/atoms/linear_model/_base.py [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+        if fit_intercept:
+            intercept_ = Variable(
+                (target_dim,),
+                name="intercept",
+                device=beta_value.device,
+                dtype=beta_value.dtype,
+            )
+            exprs["intercept"] = intercept_
+
+        super().__init__(exprs, {}, None)
```
Comment (pratikrathore8) [inline rlaopt/atoms/linear_model/_base.py]: The last argument shouldn't be `None` here, right? I believe we should specify `variable_names` to make sure we enforce that `x` and `intercept_` are indeed variables.

File: rlaopt/atoms/linear_model/_base.py [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+        defined as the proportion of deviance explained by the model:
+            
+            pseudo-R² = 1 - (model_deviance / null_deviance)
+        
+        where the null deviance is computed using a constant prediction (the mean
+        of y_true), and the model deviance uses the fitted predictions. Values
+        range from -∞ to 1, with 1 indicating perfect fit and 0 indicating the
+        model performs no better than the null model.
+        
+        Parameters
```
Comment (pratikrathore8) [inline rlaopt/atoms/linear_model/_base.py]: This docstring needs to be adjusted to be Google style.

File: rlaopt/atoms/linear_model/_inv_link_function.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    def __call__(self, linear_pred: torch.Tensor) -> torch.Tensor:
+        """Apply the inverse link function to transform linear predictor to mean parameter.
+
+        Args:
+            linear_pred: Linear predictor η = Xβ
+
+        Returns:
+            Mean parameter μ = g⁻¹(η)
+        """
+        ...
```
Comment (pratikrathore8) [inline rlaopt/atoms/linear_model/_inv_link_function.py]: Shouldn't this be `pass` rather than `...`?

File: rlaopt/atoms/linear_model/_loss_factory.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    LossType.HUBER: torch.nn.HuberLoss,
+    LossType.INV_GAUSS: InverseGaussianLoss,
+    LossType.LEAST_SQUARES: torch.nn.MSELoss,
+    LossType.LOGISTIC: torch.nn.BCEWithLogitsLoss,
+    LossType.MULTINOMIAL: torch.nn.CrossEntropyLoss,
+    LossType.POISSON: PoissonLoss,
+}
+
+
+def _get_loss_function(loss_type: LossType) -> _Loss:
```
Comment (pratikrathore8) [inline rlaopt/atoms/linear_model/_loss_factory.py]: Same thing with the underscore for this function name.

File: rlaopt/atoms/linear_model/_tweedie.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        target: Target values (y), shape (N, *)
+        power: Tweedie power parameter
+        reduction: 'none' | 'mean' | 'sum'
+        eps: Small constant for numerical stability
+
+    Returns:
+        Loss tensor (negative log-likelihood, ignoring constants)
+    """
+    # Input validation
+    if torch.any(target < 0): raise ValueError("Target values must be non-negative")
```
Comment (pratikrathore8) [inline rlaopt/atoms/linear_model/_tweedie.py]: Please put the `raise` on a separate line!

File: rlaopt/atoms/linear_model/_tweedie.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        # First term: y * ŷ^(1-p) / (1-p)
+        term1 = target * torch.pow(input_ + eps, p1) / p1
+
+        # Second term: ŷ^(2-p) / (2-p)
+        term2 = torch.pow(input_ + eps, p2) / p2
+
+        loss = -term1 + term2
+
+    # Apply reduction
+    if reduction == "none":   return loss
```
Comment (pratikrathore8) [inline rlaopt/atoms/linear_model/_tweedie.py]: The `return` statements should not be on the same line as the if statements!

File: rlaopt/atoms/linear_model/_tweedie.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        reduction: Specifies the reduction to apply to the output:
+                  'none' | 'mean' | 'sum'. Default: 'mean'
+
+    Note:
+        Predictions must be positive, targets can be zero or positive.
+        Power=1.5 is a common default, but any value in (1, 2) is valid.
+    """
+
+    def __init__(self, power: float, reduction: str = "mean"):
+        # Compound Poisson-Gamma uses power=1.5 (standard default)
```
Comment (pratikrathore8) [inline rlaopt/atoms/linear_model/_tweedie.py]: This isn't quite true for our implementation -- let's remove this comment.

File: rlaopt/utils/device_utils.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import torch
+
+
+def check_and_move_to_device(
+    tensor: torch.Tensor, device: torch.device
+) -> torch.Tensor:
+    if tensor.device != device: tensor = tensor.to(device)
```
Comment (pratikrathore8) [inline rlaopt/utils/device_utils.py]: Once again, let's separate the outcome of the `if` statement from the `if` statement itself.

File: rlaopt/atoms/linear_model/_tweedie.py (hunk 6 above)
Comment (zfrangella) [reply rlaopt/atoms/linear_model/_tweedie.py]: Same as I stated previously

**Review by pratikrathore8 — APPROVED** [review]
